### PR TITLE
[WIP] raidz expansion, alpha preview 1

### PR DESCRIPTION
**Pull request by ahrens**
### Motivation and Context

This is a alpha-quality preview of RAID-Z expansion.  This feature allows disks to be added one at a time to a RAID-Z group, expanding its capacity incrementally.  This feature is especially useful for small pools (typically with only one RAID-Z group), where there isn't sufficient hardware to add capacity by adding a whole new RAID-Z group (typically doubling the number of disks).

For additional context as well as a design overview, see my short talk from the 2017 OpenZFS Developer Summit: [slides](http://www.open-zfs.org/w/images/6/68/RAIDZ_Expansion_v2.pdf) [video](https://www.youtube.com/watch?v=ZF8V7Tc9G28)

### Description

Functionality that's currently implemented:
* Can expand raidz device with `zpool attach poolname raidz2-0 newdisk`
* Simple test script in `scripts/raidz_expand_test.sh`
* During reflow/expansion:
  * All allocated space in device is rewritten (copied to its new location in the RAIDZ vdev)
  * Reflow happens in background over multiple txg’s
  * Reads and writes during reflow are handled
  * Can reboot or export/import, resumes after import (with exception if at the very beginning of reflow)
  * Progress is reported in zpool status
* After expansion completes:
  * Can initiate additional expansions
  * Additional space available
  * Device failure and silent damage are handled
  * Can reboot or export/import
  * Status (completion time) reported in zpool status

Functionality that still needs to be implemented:
* Add on-disk feature flag
* Progress should be reported in terms of offset on disk, not bytes copied
* Logical stripe width does not increase
* Crash in the very beginning of reflow can trash the pool
* Pool must be healthy during reflow
* Does not use SIMD instructions for raidz math
* Documentation
* Automated tests!

This feature should only be used on test pools.  The pool will eventually need to be **DESTROYED**, because the on-disk format will not be compatible with the final release.  Additionally, there are currently bugs in RAID-Z expansion which can occasionally cause data loss.

I would especially appreciate if anyone has time to write some automated tests for RAID-Z expansion in the ZFS Test Suite (including converting the raidz_expand_test.sh script into a proper test).

Sponsored by: The FreeBSD Foundation

### How Has This Been Tested?

scripts/raidz_expand_test.sh

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ahrens** [comment]
cc: @emaste @allanjude @mckusick @kmoore134 @grwilson

File: module/zfs/spa.c [hunk 1]
```diff
@@ -6115,7 +6133,8 @@ spa_vdev_attach(spa_t *spa, uint64_t guid, nvlist_t *nvroot, int replacing)
 	/*
 	 * Make sure the new device is big enough.
 	 */
-	if (newvd->vdev_asize < vdev_get_min_asize(oldvd))
+	vdev_t *min_vdev = raidz ? oldvd->vdev_child[0] : oldvd;
```
Comment (kithrup) [inline module/zfs/spa.c]: This is my ignorance about the internals here, but: is that guaranteed to be the smallest device in a raidz?  It is possible to have, as an example, a raidz with four 1TB drives, and one 8TB drive.

Comment (ahrens) [reply module/zfs/spa.c]: Take a look at vdev_get_min_asize() and you'll see that the specific child doesn't actually matter.  Note that this is the same logic for replacing (attaching to) one of the disks in the raidz vdev.  In other words, for size checking purposes, `zpool attach pool raidzN-X new_disk` is the same as `zpool attach pool disk_in_raidz new_disk`.

**Comment by ahrens** [comment]
@kithrup Thanks for taking a look!  The code still has a lot of missing pieces, including a "big theory statement" of how it works.  I'm going to wait to write that until the remaining tricky pieces are implemented, so that I can be sure the comment reflects the details in the final design.

**Comment by kithrup** [comment]
No sweat.  So far, as I said, it looks good.

**Comment by ronnyegner** [comment]
@ahrens: I am willing to help (except coding). How?

**Comment by ahrens** [comment]
@ronnyegner Thanks! Things aside from coding that you could help with include:
- manual testing & reporting bugs
- writing automated tests (shell scripting)
- writing documentation
Feel free to get in touch with me directly if you'd like more specific direction.

**Comment by kwinz** [comment]
Thanks a lot @ahrens! In the talk you said that the existing old data keeps the old parity ratio aka. logical stripe width (at 6:50 and 13:20). What is the recommended way to convert the existing data to the new stripe width after expanding? Do I have to delete snapshots and `mv` data? Thanks in advance!

**Comment by ahrens** [comment]
@kwinz In the current PR, all data has the old parity ratio (even newly-written data), so there's no way to do that currently.  But before we integrate this, I'll make it so that newly-written data has the new data:parity ratio.

Then the easiest way to convert existing data would usually be to use `zfs send ... | zfs receive ...` within the same pool, and then delete the old datasets.

**Comment by mufunyo** [comment]
> Then the easiest way to convert existing data would usually be to use `zfs send ... | zfs receive ...` within the same pool, and then delete the old datasets.

That sounds like a pretty janky way to do it. It'd be a lot more user-friendly if you could just rewrite a dataset in the new data:parity ratio with a dedicated command like `zfs rewrite <dataset>` that would do it in-place.

**Comment by snajpa** [comment]
On the send/recv note, maybe @mufunyo's `zfs rewrite` could be expanded on a bit further, perhaps we could do something like KVM/similar do for live migration. I'm now wondering how many lines it would take to support live `rewrite`, which would keep on syncing until the synced delta is at minimum and then grabbing the z_teardown_lock on the original zfsvfs and swapping it with the new dataset's zfsvfs... that doesn't sound too crazy, does it?

cc @ahrens

**Comment by ronnyegner** [comment]
@ahrens I sent you two mails so far about wanting to help you. No reply. One on 6th June and another one on 8th July. Did you get them?

**Comment by ahrens** [comment]
@ronnyegner The last email I have from you is January 2015, about device removal.  I did reply to your comment on this PR on June 6th.  https://github.com/zfsonlinux/zfs/pull/8853#issuecomment-499583597

**Comment by kenthinson** [comment]
> @kwinz In the current PR, all data has the old parity ratio (even newly-written data), so there's no way to do that currently. But before we integrate this, I'll make it so that newly-written data has the new data:parity ratio.
> 
> Then the easiest way to convert existing data would usually be to use `zfs send ... | zfs receive ...` within the same pool, and then delete the old datasets.

Hello. Was wondering if a way to pause the reflow would be possible. That way we could do a workflow like this.

1. add disk paused.
2. zfs send ... | zfs receive to change stripe width. (Without io contention of the reflow)
3. delete the old data.
4. resume normal reflow (in case we as the user missed anything that needs to be moved around).

**Comment by ahrens** [comment]
@kenthinson We could add a way to pause reflow.  However, writes to locations that have not yet been reflown have to use the old stripe width.  So the workflow you outlined won't achieve your goal.  The send|recv won't change the stripe width until after the reflow has (at least mostly) completed.

**Comment by kenthinson** [comment]
@ahrens ah. Thats very interesting. I didn't think free space would have to be reflow. Thanks for the info :)

**Comment by ahrens** [comment]
@kenthinson Note that free space isn't copied, but it exists in whichever physical stripe width surrounds it.  We track how far has been reflown: whatever's before that is in the new physical stripe width.  We don't keep a map of which ranges are in which physical stripe widths.

**Comment by scineram** [comment]
Then don't you have to always use the old stripe width before reflow completes? How can stripe width be determined for a block written during reflow?

**Comment by kenthinson** [comment]
@ahrens again thanks for the information :)

**Comment by kenthinson** [comment]
@scineram that's a interesting question. I would love to understand that as well. However I'm afraid it's above my head and I don't want to bother Matt too much he's a busy guy lol.

**Comment by Matthew-Bradley** [comment]
How safe is the reflow (once everything is all tidied up) conceptually? Would the array be in a degraded state? Unavailable?

**Comment by amotin** [comment]
@ahrens Hi Matt. I've looked through the patch and I have a big worries about vdev performance  after reflow:
 - In case of logical and physical widths being different the new code generates ashift-sized I/Os to every disk to read and process each row separately.  Those will probably be aggregated later, but in case of ashift=9 the amount of CPU overhead I expect to be just huge.
 - Also, even if we try to optimize the I/Os somehow, each row now has parity data in different place(s).  That means after we expand 3-disk RAIDZ1 to 4-disk RAIDZ1 each disk on average will be able to read only about 2 ashift's of data before hitting the parity block. ZIO aggregation code will likely extend the read over the gap, but that means every disk will on average loose 50% of its linear speed.
I haven't tested this in practice yet, so please tell me that I am wrong here.

**Comment by ahrens** [comment]
@amotin Thanks for taking a look!

> In case of logical and physical widths being different the new code generates ashift-sized I/Os to every disk to read and process each row separately. Those will probably be aggregated later, but in case of ashift=9 the amount of CPU overhead I expect to be just huge.

They will definitely be aggregated later.  So we will be processing the same amount of bytes as before, but we will have more zio_t's which will cause some more overhead.  The increase in number of zio_t's would be greater with greater recordsize.  With the default recordsize=128K, and a 10-wide RAIDZ2, and 4K-sector disks, we would have around 4x as many zio_t's, because there are 4 sectors of this block on each disk.  (for 512 byte sectors it would be 32x).  I agree we'll want to measure the practical impact of this and see if there are some shortcuts we can take in the zio code to minimize the per-zio overhead for these.

> Also, even if we try to optimize the I/Os somehow, each row now has parity data in different place(s). That means after we expand 3-disk RAIDZ1 to 4-disk RAIDZ1 each disk on average will be able to read only about 2 ashift's of data before hitting the parity block. ZIO aggregation code will likely extend the read over the gap, but that means every disk will on average loose 33% of its linear speed.

Another way of thinking of this is that without expansion, each read only hits (vdev_width - parity) disks, i.e. it doesn't have to read from the "parity disks".  If you have many concurrent reads outstanding (at least vdev_width reads), you should be able to average slightly more than the max iops of a single disk, because each read leaves its parity disks idle (and different reads leave different "parity disks" idle).

After expansion, each read hits all the disks in the vdev.  In the worst case I think we'd expect (parity / old_width) reduction.

Do you have any thoughts on how we could improve this?

**Comment by amotin** [comment]
> With the default recordsize=128K, and a 10-wide RAIDZ2, and 4K-sector disks, we would have around 4x as many zio_t's, because there are 4 sectors of this block on each disk. (for 512 byte sectors it would be 32x). I agree we'll want to measure the practical impact of this and see if there are some shortcuts we can take in the zio code to minimize the per-zio overhead for these.

This worries me since even now completion of aggregated 16KB requests cause ZIO burst, stressing random code paths and locks.  Multiplying that by 32 does not sound promising to me.  It has to be tested, but not sure we have much of optimization possibilities left in ZIO pipeline, considering how much is needed.

And I would personally target this feature not to somebody having 10-wide RAIDZ2, but to 3-wide RAIDZ1, i.e. SOHO.  I think people having 10-wide RAIDZ2 is just a different market, who should not bother to add a single disk, especialy if they won't get full benefit from it.  Beside we do not recommend that wide RAIDZ's typically due to very low IOPS and long rebuild times.

> After expansion, each read hits all the disks in the vdev. In the worst case I think we'd expect (parity / old_width) reduction.

Yes. Instead of (old_width - parity) / (new_width - parity) increase in case of offload/reload.

> Do you have any thoughts on how we could improve this?

Not really. The first problem may be approached by trying to read the whole column with following data reshuffling, but that is additional memory copy -- also not great.  The problem of nonsequential data placement probably has no solution with this approach to expansion.

**Comment by justinclift** [comment]
Would a subsequent (optional?) pass to re-layout / optimise the data afterwards be feasible?

**Comment by GurliGebis** [comment]
Sorry if this has been answered already, but would this also allow "upgrading" from RAIDZ1 to RAIDZ2 and then to RAIDZ3?
My point is, lets say we have 5 drives in RAIDZ1, and wants to add 3 more - it would make pretty good sense to go to RAIDZ2 when adding one of the disks.

**Comment by Matthew-Bradley** [comment]
> Would a subsequent (optional?) pass to re-layout / optimise the data afterwards be feasible?

I'd think that would make a good default, unless you need things available post-reflow as fast as possible.

**Comment by ahrens** [comment]
@GurliGebis No, at least not in the initial design.  It would be possible to extend this work to allow new writes to use a different amount of parity.  However, it's nontrivial to change all the old data to the new parity, for similar reasons that it's hard to change all the old data to the new data:parity ratio.  You would have to allocate a different amount of space for the old blocks, so they would need to be moved, and because of snapshots, there are multiple instances of the block pointer that refers to the old location & size.

**Comment by GurliGebis** [comment]
@ahrens Would it be possible to allow reflowing the current data to use the new topology, if there is no snapshots? (Not ideal to require no snapshots, but could be a start maybe)

Sorry for asking stupid questions, but why is snapshots causing problems for changing the data to the new topology?
Wouldn't copying the data to a new location with the new topology and then change all the pointers be possible?

**Comment by Lady-Galadriel** [comment]
@GurliGebis, I think all of us would like a reflow. But, it should not be limited to RAID-Zx stripe size. That should include updating:

- checksum algorythm
- compression rate or algorythm
- encryption algorythm or master key

That would be a different discussion, probably better for the mailing list. Or a separate Issue.

**Comment by kwinz** [comment]
@Lady-Galadriel I would be happy to be able to reflow the strip size at or after expanding.
Not put this work in a separate monster project with 10 other features.

**Comment by drescherjm** [comment]
Wouldn't a reflow require block pointer rewrite?

**Comment by allanjude** [comment]
> 
> 
> Wouldn't a reflow require block pointer rewrite?

No, that is infact the point of the reflow.

We change the geometry of the RAID-Z, so if your vdev was 4 disks wide, DVA 5 which used to be on disk 1, is moved to the new disk 5, so that it remains DVA 5, and the BPs don't need to be rewritten.

see: http://zfs.datto.com/2018_slides/ahrens.pptx
and: https://www.youtube.com/watch?v=ZF8V7Tc9G28&list=PLaUVvul17xSfgoTmcgjC0MQPGC2fmljL6&index=12

**Comment by drescherjm** [comment]
I was talking about this comment and similar
> I would be happy to be able to reflow the stripe size at or after expanding so it fits the native new topology.

**Comment by thorsteneb** [comment]
Howdy, can I get some concrete guidance on how to assist, and what assistance is still needed?

I see from June 2019:
> I would especially appreciate if anyone has time to write some automated tests for RAID-Z expansion in the ZFS Test Suite (including converting the raidz_expand_test.sh script into a proper test).

I can take a stab at that, and, it might already have been done. If I am to take a stab at that, I need a good description of what "a proper test" looks like. Not the way the ZFS Test Suite works, I can figure that out, but the tasks the test would carry out.

Anything else that a hobbyist with some covid-19 imposed time and a handful of disks can do?

**Comment by ahrens** [comment]
@thorsteneb Thanks for your offer!  Those test have not been written yet, so it would be very helpful.  As far as what the tests should cover, you could start by taking a look at the (very rough) test script in `scripts/raidz_expand_test.sh`.  I'd also be happy to work with you over email, you can reach me at mahrens@delphix.com.  We can work together on brainstorming the next set of tests to write.

**Comment by ahrens** [comment]
Thank you @thorsteneb for rebasing this onto master!

**Comment by diodep** [comment]
> Thank you @thorsteneb for rebasing this onto master!

Thank you for your work to implement this feature!

So what's the status about raidz expansion now? I'm curious about how long it will take to be fully tested and merged to the mainline :D

**Comment by ahrens** [comment]
There hasn't been any implementation progress since the last update.  However, we are expecting to resume work on it this week.  Expect an update soon :-)

**Comment by ahrens** [comment]
Thank you @stuartmaybee for rebasing this onto the main branch!

**Comment by ahrens** [comment]
Thank you @fuporovvStack for getting the SIMD RAIDZ math routines working again, converting my test script to a real test case, and adding 2 additional test cases!

**Comment by GottemHams** [comment]
I noticed there has been quite some activity in the recent months. Could you give us an update on the current status and do you perhaps have a rough ETA for a somewhat usable beta version? :>

**Comment by ahrens** [comment]
@GottemHams As you noticed, I'm continuing to make progress, albeit slowly.  I'm hoping to have an updated "beta" PR in the next few weeks.

**Comment by geffao** [comment]
> 
> 
> @GottemHams As you noticed, I'm continuing to make progress, albeit slowly. I'm hoping to have an updated "beta" PR in the next few weeks.

Just passing by to say THANK YOU. Really hoping to see this feature get implemented to TrueNAS.

**Comment by kmoore134** [comment]
> > @GottemHams As you noticed, I'm continuing to make progress, albeit slowly. I'm hoping to have an updated "beta" PR in the next few weeks.
> 
> Just passing by to say THANK YOU. Really hoping to see this feature get implemented to TrueNAS.

We (On the TrueNAS side) are also watching this feature closely, you can be sure that once it lands in OpenZFS proper we'll be bringing into the product, and then doing the necessary UI/Middleware API bits to expose it.

**Comment by ahrens** [comment]
Thanks for the words of encouragement, @geffao!  I've had some time to work on this recently and have made good progress handling the last few corner cases encountered when the system crashes or a drive is damaged.

**Comment by igsol** [comment]
When this highly desired super-feature is implemented, could it be possible to expand RAIDZ1 to RAIDZ2?

**Comment by Matthew-Bradley** [comment]
> When this highly desired super-feature is implemented, could it be possible to expand RAIDZ1 to RAIDZ2?

It's been a while since this was brought up, but I'm 99.5% sure that adding another column of checksum is impossible with these changes. That would have to be an entirely different set of changes.

**Comment by geffao** [comment]
> 
> 
> > When this highly desired super-feature is implemented, could it be possible to expand RAIDZ1 to RAIDZ2?
> 
> It's been a while since this was brought up, but I'm 99.5% sure that adding another column of checksum cannot be done with this work (and is outside the scope). That would have to be an entirely different set of changes.

I'm noob in all of this, but somehow the folks at Synology have that sort of feature with their SHR-2. Would be super nice to see it happening since if a big disk goes bad, to resilver takes a long time and other disk might go bad while happening.

But hey, just wishful thinking. I'm thankful for the developments =)

**Comment by geffao** [comment]
> 
> 
> > > @GottemHams As you noticed, I'm continuing to make progress, albeit slowly. I'm hoping to have an updated "beta" PR in the next few weeks.
> > 
> > 
> > Just passing by to say THANK YOU. Really hoping to see this feature get implemented to TrueNAS.
> 
> We (On the TrueNAS side) are also watching this feature closely, you can be sure that once it lands in OpenZFS proper we'll be bringing into the product, and then doing the necessary UI/Middleware API bits to expose it.

Thank you @kmoore134 !

**Comment by Matthew-Bradley** [comment]
> > > When this highly desired super-feature is implemented, could it be possible to expand RAIDZ1 to RAIDZ2?
> > 
> > 
> > It's been a while since this was brought up, but I'm 99.5% sure that adding another column of checksum cannot be done with this work (and is outside the scope). That would have to be an entirely different set of changes.
> 
> I'm noob in all of this, but somehow the folks at Synology have that sort of feature with their SHR-2. Would be super nice to see it happening since if a big disk goes bad, to resilver takes a long time and other disk might go bad while happening.
> 
> But hey, just wishful thinking. I'm thankful for the developments =)

In such a scenario you would still have to silver the new parity disk, instead of the replacement for the one that failed.

**Comment by ahrens** [comment]
Correct, this design doesn't allow adding parity (RAIDZ1 -> RAIDZ2).  DRAID is similar to traditional RAID5/6 in that we can calculate where the data and parity are (it isn't dependent on the blocks' sizes/locations), so this would be a reasonable feature to add to DRAID (DRAID1->DRAID2, etc)

**Comment by Matthew-Bradley** [comment]
> Correct, this design doesn't allow adding parity (RAIDZ1 -> RAIDZ2). DRAID is similar to traditional RAID5/6 in that we can calculate where the data and parity are (it isn't dependent on the blocks' sizes/locations), so this would be a reasonable feature to add to DRAID (DRAID1->DRAID2, etc)

One thing occurs to me: Do these patches allow for the removal of a disk? (ie. the inverse operation, where data is re-flowed to a fewer number of disks and then one can be removed)

**Comment by ahrens** [comment]
@Matthew-Bradley No, removing a disk from RAIDZ would not be practical with this design.  Although you could technically reflow the data the same way, it wouldn't maintain the data redundancy guarantee of RAIDZ.  The problem is that most rows (i.e. one sector of parity plus the associated data) would end up with 2 sectors on the same physical disk.  So for example with RAIDZ1, we should be able to tolerate any 1 disk failing without losing data.  We can reconstruct it using the other disks' data plus the parity.  But if you do a "shrinking reflow" and then lose a disk, you could lose 2 pieces of data (or one data and the parity), and then you won't be able to reconstruct the missing data.  To diagram: (# = child disk ID; D=data; P=parity)
```
before:
0 1 2 3
D D D P
 
after:
0 1 2
D D D
P
 ```
If disk 0 is lost then we can't reconstruct the missing data, because the parity is also missing.

**Comment by Matthew-Bradley** [comment]
> @Matthew-Bradley No, removing a disk from RAIDZ would not be practical with this design. Although you could technically reflow the data the same way, it wouldn't maintain the data redundancy guarantee of RAIDZ. The problem is that most rows (i.e. one sector of parity plus the associated data) would end up with 2 sectors on the same physical disk. So for example with RAIDZ1, we should be able to tolerate any 1 disk failing without losing data. We can reconstruct it using the other disks' data plus the parity. But if you do a "shrinking reflow" and then lose a disk, you could lose 2 pieces of data (or one data and the parity), and then you won't be able to reconstruct the missing data. To diagram: (# = child disk ID; D=data; P=parity)
> 
> ```
> before:
> 0 1 2 3
> D D D P
>  
> after:
> 0 1 2
> D D D
> P
> ```
> 
> If disk 0 is lost then we can't reconstruct the missing data, because the parity is also missing.

I see. I suppose in that case you'd have no choice but to do something funny to put the parity on disk 2 and the overflowed data onto disk 0, which I guess would make it part of the next "row." That sounds like a headache, reworking what parity matches with what data.

**Comment by g-rocket** [comment]
@ahrens @Matthew-Bradley RE removing a disk.

It looks like this allows you to reflow, then change the stripe width (and rewrite data with the new wider strip width.
It *seems* like removing a disk could be handled if that could be done in the opposite order -- first reduce the stripe width (and rewrite all data), then reflow to actually remove the disk. Would probably be have some tricky edge-cases actually implement, but once we support having a strip width less than the number of drives, rewriting data with the new strip width, and reflowing data onto a different number of drives (all of which I *think* are planned as part of this feature) we shouldn't need to much in the way of new code to support removing a drive.

**Comment by ahrens** [comment]
@g-rocket That's a cool idea, however:
> rewrite all data

That's the tricky part.  Especially since you need to maintain block sharing between snapshots (& clones), and it would be nice to be able to do this with a small amount of free space (e.g. not requiring that half the space be free).

**Comment by PingRen32** [comment]
Hi, i got questions about this feature:
1. Is it possible to add multiple drives to pool at the same time, or just one at a time? such as adding 2 new drives to a existing 4 drive Z2, will it be 4-> 6 or 4 -> 5 -> 6?
2. I currently have a vdev of old SMR drives and according to current expansion method, seems SMR will suffer a lot during this process, either being adder or addee? (btw hope we could have a SMR detector/warning in the future)

**Comment by yorickdowne** [comment]
@PingRen32 As I understand it, one at a time.

Detecting SMR is really difficult because firmware may not be clear about it. The best detection is still before purchase, by looking at the list of known SMR drives.

**Comment by ahrens** [comment]
@PingRen32 I agree with what @yorickdowne wrote:
1. Expansion is one-at-a-time.  Extending this to adding multiple drives at once would be a reasonable future project.
2. RAIDZ expansion actually reads/writes sequentially, which is usually good for SMR, but it skips the free space, so it may still be hard on SMR drives.  It would be pretty easy to add a module parameter to force reading/writing of all blocks, which seems like it would be an ideal workload for SMR.  But I haven't tested this with SMR drives, and your best bet would be to avoid SMR where possible, or accept its variable performance characteristics.

**Comment by Matthew-Bradley** [comment]
> ...it would be nice to be able to do this with a small amount of free space (e.g. not requiring that half the space be free).

This seems to be a common theme with ZFS. I've seen warnings that you should always stay below 50% and definitely below 80% usage, and resilver all disks after dropping back below half-full for full performance. (due to fragmentation)

Is this an ongoing issue with ZFS? Is it mitigated at all by SSDs? Was BPRW meant to address it? I thought I'd mention this here, since this is patchset is a potential way to mitigate a full array.

**Comment by ahrens** [comment]
@Matthew-Bradley
> I've seen warnings that you should always stay below 50% and definitely below 80% usage

I don't agree with those warnings.  Lots of people successfully use ZFS all the way up to the 97% utilization limit (which is being raised so that you can get even closer to 100% utilization, see https://github.com/openzfs/zfs/pull/11023).  At work we have some of the most demanding workloads in terms of fragmentation (relational databases with recordsize=8k, compression=on, ashift=9, lots of sync (ZIL) writes with no log device) and we see reasonable performance up to 85-95% capacity.

> resilver all disks after dropping back below half-full for full performance. (due to fragmentation)

Resilvering (or scrubbing) has no impact on fragmentation, and no impact on performance once completed.

**Comment by Haravikk** [comment]
Excellent to see this feature in development!

Since it's more smaller pool oriented I'm curious though, could it be made to allow conversion of a mirror into a raidzn? I'm not familiar with whether ZFS has support for different parity algorithms under the hood, but in theory at least a two disk mirror could be considered equivalent to a two-disk raidz1 (single disk redundancy) except that the "parity" is an exact copy of the data, likewise a two-disk mirror could be considered equivalent to a degraded three-disk raidz2, or a three disk mirror equivalent to non-degraded three-disk raidz2 and so-on.

I suppose the main issue is whether ZFS can handle the old scheme using a different form of parity compared to the new one, such that conversion can happen during reflow (i.e- instead of reflowing as-is, convert the additional copies to parity blocks before writing into the new position, after checksumming)?

Granted without support for mirror conversion this feature still makes it possible to start with only a single disk worth of capacity and then expand later, but I'm thinking that in smaller setups it's a lot more common to start with a basic mirror, then want to switch, so it might be worth thinking about?

**Comment by yorickdowne** [comment]
> could it be made to allow conversion of a mirror into a raidzn

I'm pretty sure that's literally not possible in ZFS, and certainly not with this code. The good news though is that "poor person's raidz expansion" can work with any pool: 2 and 3 TB SAS disks are cheap on eBay. With a HBA, some SAS disks and 2 weeks of time for burn-in and ZFS send/recv, you can remake the pool into any setup you like. I made a 5x raidz2 into a 8x raidz2 that way.

See https://www.truenas.com/community/threads/poor-mans-raidz-expansion-also-backup-and-test-system.86434/

**Comment by CySlider** [comment]
Buy your 3rd disk, make a raidz1 with it plus a sparse file to simulate a second disk that you remove afterwards.

Copy over the the data from your mirror to the degraded raidz1 (single disk).

Remove one disk of the mirror and add it as a replacement to your raidz1 and rebuld it.

Extend your raidz1 to raidz2 with the other disk of the old mirror config.

Done.

Am 28. März 2021 13:15:01 MESZ schrieb Haravikk ***@***.***>:
>Excellent to see this feature in development!
>
>Since it's more smaller pool oriented I'm curious though, could it be
>made to allow conversion of a mirror into a raidzn? I'm not familiar
>with whether ZFS has support for different parity algorithms under the
>hood, but in theory at least a two disk mirror could be considered
>equivalent to a two-disk raidz1 (single disk redundancy) except that
>the "parity" is an exact copy of the data, likewise a two-disk mirror
>could be considered equivalent to a degraded three-disk raidz2, or a
>three disk mirror equivalent to non-degraded three-disk raidz2 and
>so-on.
>
>I suppose the main issue is whether ZFS can handle the old scheme using
>a different form of parity compared to the new one, such that
>conversion can happen during reflow (i.e- instead of reflowing as-is,
>convert the additional copies to parity blocks before writing into the
>new position, after checksumming)?
>
>Granted without support for mirror conversion this feature still makes
>it possible to start with only a single disk worth of capacity and then
>expand later, but I'm thinking that in smaller setups it's a lot more
>common to start with a basic mirror, then want to switch, so it might
>be worth thinking about?
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/openzfs/zfs/pull/8853#issuecomment-808882026

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by Matthew-Bradley** [comment]
> Extend your raidz1 to raidz2 with the other disk of the old mirror config.

This is not possible. Afaik, you cannot change the parity level of a raidz after it is created, even if you use this patchset. It only allows for the adding of a *non-parity* disk.

**Comment by CySlider** [comment]
Ah, sorry. Not RaidZ1 to RaidZ2 but RaidZ1 with 2 disks to Raidz1 with
3. Once this feature is done.

Was late. My bad.

Am 29.03.21 um 11:47 schrieb Matt:

**Comment by Haravikk** [comment]
> Buy your 3rd disk, make a raidz1 with it plus a sparse file to simulate a second disk that you remove afterwards. Copy over the the data from your mirror to the degraded raidz1 (single disk). Remove one disk of the mirror and add it as a replacement to your raidz1 and rebuld it. Extend your raidz1 to raidz2 with the other disk of the old mirror config. Done.

Thanks for the suggestion, this is actually very similar to what I've had to do in the past; in my case for a direct attached home array I only had the budget for a two disk mirror originally as I needed a Thunderbolt enclosure (and even a 4-bay ain't cheap) but what I really wanted was a 4-disk raidz2 (for the double disk redundancy as performance isn't my main concern for a backup pool), so in the end I added two disks, created a degraded raidz2 using offlined sparse files, transferred the data, swapped a disk out of the mirror and into the new pool, then repeated. Took ages, _technically_ safe (there was always two copies of the data throughout) but not super convenient.

That's why I just thought I'd ask about the possibility, as a mirror is _essentially_ equivalent to a single disk raidzn, except that its "parity" is a copy of the data. I understand the specifics probably don't allow it (or at least make it easy) though, just thought I'd ask!

**Comment by Xmister** [comment]
Sorry if it's been asked before, but I couldn't find it: Will this function work on pre-existing RAIDZ2 arrays, or only newly-provisioned ones?

I'm thinking about converting my mdadm+lvm+ext4 to ZFS because I had a few occasions of block mismatches on mdadm resyncs.
But it would only make sense for me if in a foreseeable future I will able to add new disks to it without destroying my vdevs,

**Comment by ahrens** [comment]
@Xmister 
> Will this function work on pre-existing RAIDZ2 arrays

Yes, this feature allows existing RAIDZ arrays to be expanded.

**Comment by Matthew-Bradley** [comment]
@ahrens I saw your talk at [1] as well as the one you did a year earlier on this feature.

I have some thoughts on the "transient errors become permanent" bit you talk about at the end. So, during the removal you just copy blocks by looking at the spacemap, to avoid scanning block pointers for the checksum data. It's faster, but it does leave things open to transient corruption in-flight. You mention that it might be theoretically possible to add in-flight checksum checking as an extension in future work, but the performance penalty would be similar to a scrub.

I'm admittedly a small-scale user of ZFS, but from my perspective a version of the raid-z expansion (or device removal) with this checking and similar performance to a scrub doesn't actually seem all that impactful.

Insertions or removals of a device are certainly operations that would be performed much less frequently (once?) over the lifetime of an array compared to scrubbing and so aren't too hard to accommodate. (fairly regular scrubbing is already a good practice)

If one performs a scrub before and/or after (as you mention) to try and catch preexisting damage and/or new transient errors, then one is already performing 1-2 additional scrub-like operations. If you went in the other direction and the removal/expansion operations were extended to optionally scrub the remaining data (that isn't touched) to provide similar checking coverage to a scrub, I might even be able to skip my scheduled scrub for that week. :P

[1] https://www.youtube.com/watch?v=Njt82e_3qVo

**Comment by PingRen32** [comment]
as Matthew-Bradley mentioned, would it be possible to have a option in release, like a fast and a safe mode?

Fast mode which skips checksum during process, and Safe mode that do expansion by allocating a least common multiple sized chunk buffer which got checksumed after each and redo if error occors. Though the expansion process may take longer, should still be workable. But this would bring safty back to table. Just feel like it would be great if we can somehow reduce that risk.

**Comment by ahrens** [comment]
Superseded by https://github.com/openzfs/zfs/pull/12225

**Comment by Fmstrat** [comment]
@ahrens Since this was superseded, does this mean contributors will no longer get credit for commits? (Granted, my part was the most minuscule thing on earth, but looks like there were a number in the commits)

**Comment by ahrens** [comment]
@Fmstrat The OpenZFS project generally requires PR's to be squashed to one commit, wikh multiple authorships preserved via the `Contributions-by` commit message tags.  I'm sorry I missed you and @thorsteneb!  I've added you both to the commit message and the PR description of #12225.  Let me know if there's anyone else that I missed.

**Comment by Fmstrat** [comment]
@ahrens Oh np at all, as I said, my contribution was about as small as one can be, I just remembered seeing other names in there when I submitted. Thanks! And thanks for all the work on this feature, so happy to see it come out of alpha.

**Comment by justinclift** [comment]
Yep, I think half the people in the storage world are going to be happy when this is released (and production ready moreso!). :smile:

**Pull request closed without merging**